### PR TITLE
Add streaming order book support for CEX.IO

### PR DESCRIPTION
**Pull request by BenjaminTrapani**
The code still needs to be formatted per mvn com.coveo:fmt-maven-plugin:format. The original source was not formatted per this extension, so let's revert commit 9eb03e03f2c1cc3df8c33b9ee5e125adfa7bfd1f right before merging to restore the correct format.

### CODE REVIEW

File: xchange-cexio/src/main/java/info/bitrich/xchangestream/cexio/dto/CexioWebSocketOrderBookSubscriptionData.java [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+	final boolean subscribe;
+	
+	@JsonProperty
+	final int depth;
+
+	public CexioWebSocketOrderBookSubscriptionData(CurrencyPair currencyPair, boolean subscribe) {
+		pair = new ArrayList<String>(Arrays.asList(currencyPair.base.toString(), 
+				currencyPair.counter.toString()));
+		this.subscribe = subscribe;
+		this.depth = 10; //-1 for full depth
```
Comment (BenjaminTrapani) [inline xchange-cexio/src/main/java/info/bitrich/xchangestream/cexio/dto/CexioWebSocketOrderBookSubscriptionData.java]: This might be worth exposing at the API layer. Not sure what is the best way to do so because the xchange API does not provide a way of doing this now.

File: xchange-cexio/src/main/java/info/bitrich/xchangestream/cexio/CexioStreamingMarketDataService.java [hunk 2]
```diff
@@ -1,18 +1,74 @@
...
+        	}
+			return retVal;
+		}
+    }
+    
+	@Override
     public Observable<OrderBook> getOrderBook(CurrencyPair currencyPair, Object... args) {
-        throw new NotYetImplementedForExchangeException();
+		
+		streamingOrderDataService.waitForAuthIfNeeded();
```
Comment (badgerwithagun) [inline xchange-cexio/src/main/java/info/bitrich/xchangestream/cexio/CexioStreamingMarketDataService.java]: Why are you authenticating?  Most exchanges don't need it for public data like order books.

Might be worth having a read of #274 to make sure this is heading in the same direction (it looks like it is, at a glance)

File: xchange-cexio/src/main/java/info/bitrich/xchangestream/cexio/CexioStreamingRawService.java [hunk 3]
```diff
@@ -23,44 +30,139 @@
...
         super(apiUrl, Integer.MAX_VALUE);
+        this.isTest = isTest;
     }
-
+    
+    public void waitForAuthIfNeeded() {
+    	authLock.lock();
+    	try
+    	{
+    		while (!isAuthenticated && !isTest)
```
Comment (badgerwithagun) [inline xchange-cexio/src/main/java/info/bitrich/xchangestream/cexio/CexioStreamingRawService.java]: Could you remove the `isTest` argument and simply use `apiKey != null`?  That is, authenticate if you have authentication details, and don't authenticate if you don't have them (matches other exchanges - see #274)

File: xchange-cexio/src/main/java/info/bitrich/xchangestream/cexio/CexioStreamingRawService.java [hunk 4]
```diff
@@ -23,44 +30,139 @@
...
+			// TODO Auto-generated catch block
+			e.printStackTrace();
+		}
+    	finally
+    	{
+    		authLock.unlock();
+    	}
+    }
+    
+    public void signalAuthComplete() {
```
Comment (badgerwithagun) [inline xchange-cexio/src/main/java/info/bitrich/xchangestream/cexio/CexioStreamingRawService.java]: This lock processing would be much more cleanly implemented with a `CountdownLatch`.

File: xchange-cexio/src/main/java/info/bitrich/xchangestream/cexio/CexioAdapters.java [hunk 5]
```diff
@@ -71,4 +78,53 @@ private static BigDecimal adaptAmount(BigDecimal amount) {
         }
         return status;
     }
+    
+    private static List<LimitOrder> updateOrInsertQuantityForPrice(List<LimitOrder> existingOrders,
```
Comment (badgerwithagun) [inline xchange-cexio/src/main/java/info/bitrich/xchangestream/cexio/CexioAdapters.java]: Apologies if I've not read this in enough detail, but are you aware of all the `update` methods on `OrderBook` which are intended to do this sort of thing?  Is this custom logic definitely necessary?

In some cases it is; some exchanges are weird.  Just want to check.

File: xchange-cexio/src/main/java/info/bitrich/xchangestream/cexio/CexioAdapters.java [hunk 6]
```diff
@@ -71,4 +78,53 @@ private static BigDecimal adaptAmount(BigDecimal amount) {
...
+    		} 
+    		else 
+    		{
+    			existingOrders.add(insertIndex, orderForUpdatedQuantityPrice);
+    		}
+    	}
+    	return existingOrders;
+    }
+    
+    protected static CurrencyPair adaptCurrencyPairString(String currencyPairString) {
```
Comment (badgerwithagun) [inline xchange-cexio/src/main/java/info/bitrich/xchangestream/cexio/CexioAdapters.java]: I'd be very surprised if XChange doesn't already provide a method for this!

File: xchange-cexio/src/main/java/info/bitrich/xchangestream/cexio/CexioStreamingRawService.java (hunk 4 above)
Comment (badgerwithagun) [reply xchange-cexio/src/main/java/info/bitrich/xchangestream/cexio/CexioStreamingRawService.java]: There's another issue here that there is no escape from this lock in the event of a disconnect.  The usual pattern is something like this:

```
while (latch.await(5, SECONDS)) {
  if (disconnected)
    break;
}
```
I wonder, though, if this would be better implemented by performing the auth negotiation _before_ returning complete in the `Completable` returned by `connect()`.  That way you can be sure that this handshaking process will definitely have completed before the caller starts asking for subscriptions.

I suspect a few other exchanges are vulnerable to this issue (particularly Bitfinex, which I'm working on at the moment), so thank you for highlighting it!

File: xchange-cexio/src/main/java/info/bitrich/xchangestream/cexio/CexioStreamingMarketDataService.java (hunk 2 above)
Comment (BenjaminTrapani) [reply xchange-cexio/src/main/java/info/bitrich/xchangestream/cexio/CexioStreamingMarketDataService.java]: For some reason Cex.io requires users to authenticate before gaining access to the order book. This change seems to follow the logic outlined in #274 so far as elevating this stream to an authenticated one, but the mechanism isn't very clean. Since all Cex.IO APIs require authentication, I'll use the connect completable to signal that the authentication has completed.

File: xchange-cexio/src/main/java/info/bitrich/xchangestream/cexio/CexioStreamingRawService.java (hunk 3 above)
Comment (BenjaminTrapani) [reply xchange-cexio/src/main/java/info/bitrich/xchangestream/cexio/CexioStreamingRawService.java]: Done in d1c7ab91daaa8867f279993a72e812c2e3dd4e90

Comment (BenjaminTrapani) [reply xchange-cexio/src/main/java/info/bitrich/xchangestream/cexio/CexioStreamingRawService.java]: We do not need an isTest argument for the existing test suite because none of the tests connect to the exchange, so bypass waiting on the returned completable.

File: xchange-cexio/src/main/java/info/bitrich/xchangestream/cexio/CexioStreamingRawService.java (hunk 4 above)
Comment (BenjaminTrapani) [reply xchange-cexio/src/main/java/info/bitrich/xchangestream/cexio/CexioStreamingRawService.java]: The Completable abstract class provides blocking await functions and an overload for a timeout. Callers can wait on the connect for a specific duration before deciding that the connect failed. We could also trigger an error on the completable on auth failures, which might be better. Should I add code to invoke OnError on auth failures?

File: xchange-cexio/src/main/java/info/bitrich/xchangestream/cexio/CexioAdapters.java (hunk 5 above)
Comment (BenjaminTrapani) [reply xchange-cexio/src/main/java/info/bitrich/xchangestream/cexio/CexioAdapters.java]: OrderBook update did exactly what I needed. This is fixed in d1c7ab91daaa8867f279993a72e812c2e3dd4e90

File: xchange-cexio/src/main/java/info/bitrich/xchangestream/cexio/CexioAdapters.java (hunk 6 above)
Comment (BenjaminTrapani) [reply xchange-cexio/src/main/java/info/bitrich/xchangestream/cexio/CexioAdapters.java]: They do but it is private :( I copied their implementation exactly with a note in the code about the duplication. I can submit a PR to them to make the function protected if that would be useful.

**Comment by BenjaminTrapani** [comment]
@badgerwithagun These changes should be ready to go now. I reformatted xchange-cexio per the mvn format used in xchange in the final commit, so it is easiest to read the updates by looking at the diffs for the individual commits before 3e16614.

**Comment by BenjaminTrapani** [comment]
@badgerwithagun Ping on this one. Is there anything I missed in the feedback? How does the most recent iteration look? If it looks alright, I'll sync up with latest develop.

**Comment by badgerwithagun** [comment]
Hey @BenjaminTrapani, sorry, I've been busy. Will look at this ASAP.

**Comment by badgerwithagun** [comment]
OK, sorry about the delay  @BenjaminTrapani . It's clear there's no-one else using CEX.IO, so I'm just going to take a leaf out of the XChange project's book and merge this.  It can't hurt if only you are affected!
